pennylane_lightning_gpu/src/simulator/MPIWorker.hpp

```diff
@@ -0,0 +1,273 @@
+        PL_CUDA_IS_SUCCESS(cudaFree(d_transferWorkspace));
+        for (auto *d_subSV : d_subSVsP2P)
+            PL_CUDA_IS_SUCCESS(cudaIpcCloseMemHandle(d_subSV));
+        for (auto event : remoteEvents)
+            PL_CUDA_IS_SUCCESS(cudaEventDestroy(event));
+        PL_CUDA_IS_SUCCESS(cudaEventDestroy(localEvent));
+    }
+};
+
+using SharedLocalStream =
+    std::shared_ptr<std::remove_pointer<cudaStream_t>::type>;
+using SharedMPIWorker = std::shared_ptr<
+    std::remove_pointer<custatevecSVSwapWorkerDescriptor_t>::type>;
+
+/**
+ * @brief Creates a SharedLocalStream (a shared pointer to a cuda stream)
+ */
+inline SharedLocalStream make_shared_local_stream() {
+    cudaStream_t localStream;
+    PL_CUDA_IS_SUCCESS(cudaStreamCreate(&localStream));
+    return {localStream, MPIWorkerDeleter()};
+}
+
+/**
+ * @brief Creates a SharedMPIWorker (a shared pointer to a
+ * custatevecSVSwapWorker)
+ *
+ * @param handle custatevecHandle.
+ * @param mpi_manager MPI manager object.
+ * @param sv Pointer to the data requires MPI operation.
+ * @param numLocalQubits Number of local qubits.
+ * @param localStream Local cuda stream.
+ */
+
+template <typename CFP_t>
+inline SharedMPIWorker make_shared_mpi_worker(custatevecHandle_t handle,
+                                              MPIManager &mpi_manager,
+                                              CFP_t *sv, size_t numLocalQubits,
+                                              cudaStream_t localStream) {
+
+    custatevecSVSwapWorkerDescriptor_t svSegSwapWorker = nullptr;
+
+    int nDevices_int = 0;
+    PL_CUDA_IS_SUCCESS(cudaGetDeviceCount(&nDevices_int));
+
+    size_t nDevices = static_cast<size_t>(nDevices_int);
+
+    // Ensure the P2P devices is calulcated based on the number of MPI processes
+    // within the node
+    nDevices = mpi_manager.getSizeNode() < nDevices ? mpi_manager.getSizeNode()
+                                                    : nDevices;
+
+    size_t nP2PDeviceBits =
+        std::bit_width(static_cast<unsigned int>(nDevices)) - 1;
+
+    cudaDataType_t svDataType;
+    if constexpr (std::is_same_v<CFP_t, cuDoubleComplex> ||
+                  std::is_same_v<CFP_t, double2>) {
+        svDataType = CUDA_C_64F;
+    } else {
+        svDataType = CUDA_C_32F;
+    }
+
+    cudaEvent_t localEvent = nullptr;
+    custatevecCommunicatorDescriptor_t communicator = nullptr;
+
+    PL_CUDA_IS_SUCCESS(cudaEventCreateWithFlags(
+        &localEvent, cudaEventInterprocess | cudaEventDisableTiming));
+
+    custatevecCommunicatorType_t communicatorType;
+    if (mpi_manager.getVendor() == "MPICH") {
+        communicatorType = CUSTATEVEC_COMMUNICATOR_TYPE_MPICH;
+    }
+    if (mpi_manager.getVendor() == "Open MPI") {
+        communicatorType = CUSTATEVEC_COMMUNICATOR_TYPE_OPENMPI;
+    }
+
+    const char *soname0 = nullptr;
+    auto err = custatevecCommunicatorCreate(handle, &communicator,
+                                            communicatorType, soname0);
```

Comment:
I wonder if we need to set this variable just to use it in the next line.
Would something like that work?
```suggestion
    auto err = custatevecCommunicatorCreate(handle, &communicator,
                                            communicatorType, nullptr);
```

Reply:
Thanks @AmintorDusko! Good point! Will check if it works after Perlmutter is back.

Reply:
Works on Perlmutter & just updated it!